python/pyarrow/_fs.pyx

```diff
@@ -0,0 +1,292 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# cython: language_level = 3
+
+from pathlib import Path
+
+from pyarrow.compat import frombytes, tobytes
+from pyarrow.includes.common cimport *
+from pyarrow.includes.libarrow_fs cimport *
+from pyarrow.lib import _detect_compression
+from pyarrow.lib cimport (
+    check_status,
+    NativeFile,
+    BufferedOutputStream,
+    BufferedInputStream,
+    CompressedInputStream,
+    CompressedOutputStream
+)
+
+
+cdef inline c_string _to_string(p):
+    # supports types: byte, str, pathlib.Path
+    return tobytes(str(p))
+
+
+cpdef enum FileType:
+    NonExistent
+    Uknown
+    File
+    Directory
+
+
+cdef class TimePoint:
+    pass
+
+
+cdef class FileStats:
+
+    cdef CFileStats stats
+
+    def __init__(self):
+        raise TypeError('dont initialize me')
+
+    @staticmethod
+    cdef FileStats wrap(CFileStats stats):
+        cdef FileStats self = FileStats.__new__(FileStats)
+        self.stats = stats
+        return self
+
+    @property
+    def type(self):
+        cdef CFileType ctype = self.stats.type()
+        if ctype == CFileType_NonExistent:
+            return FileType.NonExistent
+        elif ctype == CFileType_Unknown:
+            return FileType.Uknown
+        elif ctype == CFileType_File:
+            return FileType.File
+        elif ctype == CFileType_Directory:
+            return FileType.Directory
+        else:
+            raise ValueError('Unhandled FileType {}'.format(type))
+
+    @property
+    def path(self):
+        return Path(frombytes(self.stats.path()))
+
+    @property
+    def base_name(self):
+        return frombytes(self.stats.base_name())
+
+    @property
+    def size(self):
+        return self.stats.size()
+
+    @property
+    def extension(self):
+        return frombytes(self.stats.extension())
+
+    # @property
+    # def mtime(self):
+    #     return self.stats.mtime()
```

Comment:
Well, it would be interesting to get this right (meaning, probably return a `datetime`?).